libraries/node-core-library/src/Terminal/Terminal.ts

```diff
@@ -0,0 +1,281 @@
+
+/**
+ * This class facilitates writing to a console.
+ *
+ * @beta
+ */
+export class Terminal {
+  public verboseEnabled: boolean;
+
+  private _providers: Set<ITerminalProvider>;
+
+  public constructor(provider: ITerminalProvider, verboseEnabled: boolean = false) {
+    this._providers = new Set<ITerminalProvider>();
+    this._providers.add(provider);
+
+    this.verboseEnabled = verboseEnabled;
+  }
+
+  public registerProvider(provider: ITerminalProvider): void {
+    this._providers.add(provider);
+  }
+
+  public unregisterProvider(provider: ITerminalProvider): void {
+    if (this._providers.has(provider)) {
+      this._providers.delete(provider);
+    }
+  }
+
+  /**
+   * Write a generic message to the terminal
+   */
+  public write(...messageParts: (string | IColorableSequence)[]): void {
+    this._writeSegmentsToProviders(messageParts, Severity.log);
+  }
+
+  /**
+   * Write a generic message to the terminal, followed by a newline
+   */
+  public writeLine(...messageParts: (string | IColorableSequence)[]): void {
+    this.write(...messageParts, { text: EOL });
+  }
+
+  /**
+   * Write a warning message to the console with yellow text.
+   */
+  public writeWarning(...messageParts: (string | IColorableSequence)[]): void {
+    this._writeSegmentsToProviders(
+      messageParts.map(
+        (part) => ({ ...(typeof part === 'string' ? { text: part } : part), foregroundColor: ColorValue.Yellow })
+      ),
+      Severity.warn
+    );
+  }
+
+  /**
+   * Write a warning message to the console with yellow text, followed by a newline.
+   */
+  public writeWarningLine(...messageParts: (string | IColorableSequence)[]): void {
+    this._writeSegmentsToProviders(
+      [
+        ...messageParts.map(
+          (part) => ({ ...(typeof part === 'string' ? { text: part } : part), foregroundColor: ColorValue.Yellow })
+        ),
+        { text: EOL }
+      ],
+      Severity.warn
+    );
+  }
+
+  /**
+   * Write an error message to the console with red text.
+   */
+  public writeError(...messageParts: (string | IColorableSequence)[]): void {
+    this._writeSegmentsToProviders(
+      messageParts.map(
+        (part) => ({ ...(typeof part === 'string' ? { text: part } : part), foregroundColor: ColorValue.Red })
+      ),
+      Severity.error
+    );
+  }
+
+  /**
+   * Write an error message to the console with red text, followed by a newline.
+   */
+  public writeErrorLine(...messageParts: (string | IColorableSequence)[]): void {
+    this._writeSegmentsToProviders(
+      [
+        ...messageParts.map(
+          (part) => ({ ...(typeof part === 'string' ? { text: part } : part), foregroundColor: ColorValue.Yellow })
+        ),
+        { text: EOL }
+      ],
+      Severity.error
+    );
+  }
+
+  /**
+   * Write a verbose-level message. Messages are only written if verbose mode is turned on.
+   */
+  public writeVerbose(...messageParts: (string | IColorableSequence)[]): void {
+    if (this.verboseEnabled) {
```

Comment:
>if (this.verboseEnabled) { [](start = 4, length = 26)

Perhaps this should be determined by the provider.  I can imagine writing verbose messages to a log file, but not to the console.